### PR TITLE
Gradle build improvements

### PR DESCRIPTION
**Pull request by tpreviero**
This PR provide little improvements to the Gradle build. 

## Description
Here's a brief list of what changed:
- Add com.google.protobuf:protoc dependency for configurations
- Update version of Gradle Enterprise plugin to version 3.12.4
- Enable local build caching
- Use mavenCentral instead of jcenter as repository
- Use archiveFileName instead of deprecated archiveName property
- Extract swagger download from copySwagger task Explicitly define inputs and outputs for copyGWActions task
- Add normalization to ignore property git.build.time in git.properties files in order to cache some task execution

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components
- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Scheduler
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [x] General tooling
- [ ] Documentation

## Types of changes
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md#coding-standards) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: core/scheduler/build.gradle
Comment (jprinet) [inline core/scheduler/build.gradle]: I don't understand how the project can be built if a dependency is missing 🤷‍♂️

File: core/standalone/build.gradle
Comment (jprinet) [inline core/standalone/build.gradle:42]: Could you check that the download is not happening if the file is present?

File: core/standalone/build.gradle
Comment (jprinet) [inline core/standalone/build.gradle]: I would move that into the task definition

File: core/standalone/build.gradle
Comment (jprinet) [inline core/standalone/build.gradle]: This comment is not clear to me

**Review by jprinet — CHANGES_REQUESTED** [review]
Let's have @runningcode to review this as well

File: core/standalone/build.gradle
Comment (tpreviero) [reply core/standalone/build.gradle:42]: Yes, that appear to work properly.
Running a `clean build -x test` followed by a `build` doesn't download anything.

File: core/standalone/build.gradle
Comment (tommaso-previero) [reply core/standalone/build.gradle:148]: It should be done with 80959f9832cce44ad64de3bee2a391f00a907e9e

File: core/standalone/build.gradle
Comment (tpreviero) [reply core/standalone/build.gradle]: It should be done with 80959f9832cce44ad64de3bee2a391f00a907e9e

File: core/standalone/build.gradle
Comment (tpreviero) [reply core/standalone/build.gradle]: What would you think of

>  Ignores the property `git.build.time` in `git.properties` files to allow `compileScala` task to be cached
 
Is it more clear?

File: core/scheduler/build.gradle
Comment (tpreviero) [reply core/scheduler/build.gradle]: I honestly don't understand that either.
I wonder if it's something related to to Operating System where the build is running on.
The error you get when removing that dependency is the following:
```
> Task :core:scheduler:generateProto FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':core:scheduler:generateProto'.
> Could not resolve all files for configuration ':core:scheduler:protobufToolsLocator_protoc'.
   > Could not find protoc-3.4.0-osx-aarch_64.exe (com.google.protobuf:protoc:3.4.0).
     Searched in the following locations:
         https://repo.maven.apache.org/maven2/com/google/protobuf/protoc/3.4.0/protoc-3.4.0-osx-aarch_64.exe
```
Not sure what to do to be honest.

Comment (jprinet) [reply core/scheduler/build.gradle]: Alright, now makes sense, there are some `osx` compatible candidates in there with different _arch_:
https://repo.maven.apache.org/maven2/com/google/protobuf/protoc/3.4.0/

This one should probably be handled first with an issue

**Comment by tpreviero** [comment]
That makes sense. Do you think we could still address it within the same PR?

**Comment by jprinet** [comment]
You may try to fetch an artifact matching your arch with some extra attributes in the `dependency` instruction. I still would address that with an issue first as this could be on purpose.

File: core/scheduler/build.gradle
Comment (tpreviero) [reply core/scheduler/build.gradle]: I see, trying building the project within a Linux container the problem persists. 
The dependency seems needed regardless the actual operating system. I'm not fully getting it to be honest.
Shall I open already the issue on the original project? Should I wait for the final review?

Comment (jprinet) [reply core/scheduler/build.gradle]: I think this problem appeared with M1 architecture. You can probably offer a solution within your PR with inspiration taken from [there](https://github.com/grpc/grpc-java/issues/7690)

Comment (tpreviero) [reply core/scheduler/build.gradle]: Or I can upgrade the `protoc` library, as already did, to a version >= `3.18.0` that include a version for the new architectures.
Which one do you reckon would be better?

Comment (jprinet) [reply core/scheduler/build.gradle]: If a straight update solves it, I'd rather take this approach

Comment (tpreviero) [reply core/scheduler/build.gradle]: I've been digging a little into the issue (Build Scans™️ are awesome 🚀 ).
Apparently the version of `protoc` is brought into the build with the plugin `com.lightbend.akka.grpc.gradle` of which the version is `1.0.0`. There are [more recent versions of that plugin](https://plugins.gradle.org/plugin/com.lightbend.akka.grpc.gradle), yet upgrading its version bring others compilation issues. 

Being `3.18.0` the first version of `com.google.protobuf:protoc` containing the required files for M1, I downgraded the version used to that particular version.
I also tried with a more descriptive approach substituting the dependency in eb7485c2c1533b461e911e05f2de356709ff3560.

Open to suggestions 🙃

File: core/scheduler/build.gradle
Comment (runningcode) [inline core/scheduler/build.gradle:69]: I'm curious why do we use the `dependencySubstitution` approach here instead of the `force` approach used by all the other versions above?

File: core/standalone/build.gradle
Comment (runningcode) [inline core/standalone/build.gradle:58]: just a word of caution:
in general we shouldn't set `project.ext` variables like this inside a task configuration action. if the task is migrated to use the task configuration avoidance API then this variable would not be set.

That being said, I think we should keep it like this for now due to the fact that we don't want to create a massive diff.

File: core/standalone/build.gradle
Comment (runningcode) [inline core/standalone/build.gradle:41]: another thought, there is likely a swagger Gradle plugin that will do this work for you. no need to migrate for the purposes of this PR, im just writing it here for learning.

File: core/standalone/build.gradle
Comment (runningcode) [inline core/standalone/build.gradle]: The task will be cached whether or not this property is ignored. Here's a better wording.
```suggestion
// Ignores the property `git.build.time` in `git.properties` files to allow `compileScala` task to be re-used from the cache when the timestamp changes.
```

That being said, the compileScala task should not use runtimeClasspath as an input. Is it doing so? Can you share the comparison?

File: settings.gradle
Comment (runningcode) [inline settings.gradle:42]: I would remove line 30-35 since they are redundant.

**Review by runningcode — COMMENTED** [review]
Added some comments mostly for learning. Thanks for creating this PR.
Let's not make the PR to big though since this is an open source project and we want to make a small diff.

File: core/scheduler/build.gradle
Comment (tpreviero) [reply core/scheduler/build.gradle:69]: Originally I was using the `force` approach, which also works fine, but then I thought that this approach looks more "idiomatic". 
I can put back the `force` if you think that it adds less cognitive load.

File: core/standalone/build.gradle
Comment (tpreviero) [reply core/standalone/build.gradle:58]: Just to double-check, do you suggest to remove the introduction of `project.ext` and to use a local variable instead?
That variable would need to be shared between `downloadSwagger` and `copySwagger` tasks.
Extracting the download part from the original task has as goal avoiding to always download the archive.
I'm assume we keep that part, right?

File: core/standalone/build.gradle
Comment (tpreviero) [reply core/standalone/build.gradle]: I've been digging a little into the issue since I couldn't replicate it.
Apparently updating the `com.gorylenko.gradle-git-properties` to the `2.4.1` version caused the property `git.build.time` not to be included anymore into the `git.properties`, so the normalization is not needed anymore, I'll remove it.
Anyhow, just for the sake of understanding things, I tried rolling back the version of the plugin to the `2.0.0` and the task whose cache couldn't be reused was the `core:standalone:compileTestScala` and not the `compileScala` as I mistakenly said in the comment.
That said, I'll remove the normalization altogether.

Comment (tpreviero) [reply core/standalone/build.gradle]: Removed with 67447e90339eb2ff18a9092c2fd2051e23a0d310

File: settings.gradle
Comment (tpreviero) [reply settings.gradle:42]: Removed with 67447e90339eb2ff18a9092c2fd2051e23a0d310

File: core/scheduler/build.gradle
Comment (runningcode) [reply core/scheduler/build.gradle:69]: I will leave this to the project maintainers to decide.

File: core/standalone/build.gradle
Comment (runningcode) [reply core/standalone/build.gradle:58]: Yes, using a local variable would be the best solution here but let's not refactor all of their build logic in a single PR.

**Review by runningcode — APPROVED** [review]
Thanks. Please open this PR against the main project.

File: .github/workflows/0-on-demand.yaml
Comment (runningcode) [inline .github/workflows/0-on-demand.yaml:67]: Why are we changing the github actions here?

Comment (tpreviero) [reply .github/workflows/0-on-demand.yaml:67]: I honestly missed this change and I don't really remember 😅
The [actual PR opened to the project](https://github.com/apache/openwhisk/pull/5387/files) didn't have this change.
Sorry for the confusion, this was just a "toy PR" for our discussions, I'll close it.

**Pull request closed without merging**
